packages/react-components/src/components/Input/AutoComplete.tsx

```diff
@@ -0,0 +1,250 @@
+
+const buildOptionsFromAutoCompleteListItems = (
+  options: IAutoCompleteListItem[]
+): IPickerListItem[] =>
+  options.map(({ customElement, name, ...rest }) => ({
+    name,
+    key: name,
+    ...(customElement && {
+      customElement: {
+        listItemBody: customElement,
+        selectedItemBody: <></>,
+      },
+    }),
+    ...rest,
+  }));
+
+const getFilteredPickerItems = (
+  items: IPickerListItem[],
+  single: boolean,
+  hideIfExactMatch: boolean,
+  inputValue: string
+): IPickerListItem[] => {
+  const isExactMatch = items.length === 1 && items[0].name === inputValue;
+  const shownItems = single ? items.slice(0, 1) : items;
+
+  return hideIfExactMatch && isExactMatch ? [] : shownItems;
+};
+
+/**
+ * Text input with autocomplete functionality. The autocomplete list is displayed below the input and is accessible via the keyboard
+ * and mouse. The list is filtered based on the input value.
+ *
+ * @example
+ * <AutoComplete
+ *  options={['Paul', 'Adam', 'John']}
+ *  placeholder="Type a name"
+ * />
+ */
+export const AutoComplete = React.forwardRef<
+  HTMLInputElement,
+  AutoCompleteProps
+>(
+  (
+    {
+      disabled,
+      minListHeight = MIN_LIST_HEIGHT,
+      maxListHeight = DEFAULT_LIST_HEIGHT,
+      autocompleteOpenOnInit = false,
+      placement,
+      options,
+      readOnly,
+      floatingStrategy,
+      single,
+      alwaysShowAllOptions,
+      hideIfExactMatch = true,
+      ...inputProps
+    },
+    ref
+  ) => {
+    const inputRef = React.useRef<HTMLInputElement>(null);
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    React.useImperativeHandle(ref, () => inputRef.current!, []);
+
+    const [value, setValue] = React.useState('');
+    const isValueControlled = inputProps.value !== undefined;
+    const inputValue = isValueControlled ? inputProps.value : value;
+
+    const [isPickerOpen, setIsPickerOpen] = React.useState(
+      autocompleteOpenOnInit
+    );
+    const handleVisibilityChange = (newValue: boolean) => {
+      setIsPickerOpen(newValue);
+    };
+
+    const onFilter = (value: string) => {
+      handleOnFilter(value);
+    };
+
+    const handleValueChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+      inputProps.onChange?.(event);
+      setValue(event.target.value);
+      if (!alwaysShowAllOptions) {
+        onFilter(event.target.value);
+      }
+      inputRef.current?.focus();
+    };
+
+    const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+      if (virtualItemRef.current?.id && event.key === 'Enter') {
+        handleAutoComplete([{ key: virtualItemRef.current?.id }]);
+      }
+      inputProps.onKeyDown?.(event);
+    };
+
+    const handleAutoComplete = (value: { key: string }[] | null) => {
+      if (value) {
+        const key = value[0].key;
+        handleValueChange({
+          target: { value: key },
+        } as React.ChangeEvent<HTMLInputElement>);
```

Comment:
Do we need assertion here? 🤔

Reply:
Yes, otherwise it complains about a lack of 
```
addEventListener
dispatchEvent
removeEventListener
```